### PR TITLE
MTM-41872: Define Codec API definition for implement a codec ms.

### PR DESCRIPTION
**Pull request by atulpandasag**
Initial commit for the LPWAN SDK which includes the API class and interface definitions.

### CODE REVIEW

File: c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/CodecMicroservice.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package com.cumulocity.lpwan.codec.sdk;
+
+import com.cumulocity.lpwan.codec.sdk.model.DeviceInfo;
+import org.springframework.stereotype.Component;
+
+import java.util.Set;
+
+/**
+ * This abstract class will be implemented to provide a set of unique supported devices.
```
Comment (bbyreddy) [inline c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/CodecMicroservice.java]: <h1>Codec Microservice</h1>
CodecMicroservice should be implemented by the microservice implementer to advertise the devices it supports.

**Review by bbyreddy — COMMENTED** [review]
Add copyright text to all the files, including the POM file.

**Review by bbyreddy — COMMENTED** [review]
1. I thought we decided to remove the literal SDK from the package name.
2. Move the CodecMicroserviceApplication annotation into the annotation package.

File: c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/CodecMicroserviceApplication.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import com.cumulocity.lpwan.codec.sdk.service.CodecService;
+import com.cumulocity.microservice.autoconfigure.MicroserviceApplication;
+import org.springframework.context.annotation.Import;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target({ElementType.TYPE})
```
Comment (bbyreddy) [inline c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/CodecMicroserviceApplication.java]: - [ ] Add a comment to explain the purpose of this annotation and where the microservice implementer should be using it.

File: c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/Decoder.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package com.cumulocity.lpwan.codec.sdk;
+
+import com.cumulocity.lpwan.codec.sdk.exception.DecoderException;
+import com.cumulocity.lpwan.codec.sdk.model.DecodeResponse;
+import com.cumulocity.lpwan.codec.sdk.model.DeviceInfo;
+import org.springframework.stereotype.Component;
+
+/**
+ * This interface will be implemented to provide a decoding logic.
```
Comment (bbyreddy) [inline c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/Decoder.java]: - [ ] <h1>Decode data received from a device</h1>
Decoder should be implemented to provide the decoding logic.

File: c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/Decoder.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ */
+@Component
+public interface Decoder {
+    /**
+     * This method should take the following params and should return the decoded response with mapping.
```
Comment (bbyreddy) [inline c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/Decoder.java]: - [ ] This method should implement the logic to decode the data received from the device.
- [ ] describe the parameters payload, deviceId, deviceInfo and return class name and description.

File: c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/model/DecodeResponse.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+import java.util.List;
+
+@Data
+public class DecodeResponse {
+    private List<MeasurementRepresentation> measurementsToCreate;
+    private List<EventRepresentation> eventsToCreate;
+    private List<AlarmRepresentation> alarmsToCreate;
+    private List<String> alarmTypesToClear;
+    private List<ManagedObjectProperty> propertiesToAdd;
```
Comment (bbyreddy) [inline c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/model/DecodeResponse.java]: - [ ] propertiesToAddToDeviceManagedObject

File: c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/model/DeviceInfo.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package com.cumulocity.lpwan.codec.sdk.model;
```
Comment (bbyreddy) [inline c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/model/DeviceInfo.java]: Add Java doc comments

File: c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/model/ManagedObjectProperty.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package com.cumulocity.lpwan.codec.sdk.model;
```
Comment (bbyreddy) [inline c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/model/ManagedObjectProperty.java]: - [ ] Java doc comment

File: c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/service/CodecService.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+@Service
+@Slf4j
+public class CodecService {
+
+    @Autowired
+    private Decoder decoder;
+
+    /**
+     * This method should decode the payload received by a particular device.
```
Comment (bbyreddy) [inline c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/service/CodecService.java]: - [ ] This method invokes the decode logic provided by the microservice implementer by implementing the Decoder interface.

File: c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/service/CodecService.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+     * @param decode
+     * @return DecodedData
+     */
+    public DecodeResponse decode(DecodeRequest decode) throws DecoderException {
+        log.debug("Forwarding decoding request for the device with Id {} with payload {}", decode.getDeviceId(), decode.getPayload());
+        return decoder.decode(null, null, null);
+    }
+
+    /**
+     * This method should register a device type upon subscribing the codec microservice.
```
Comment (bbyreddy) [inline c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/service/CodecService.java]: - [ ] This method creates a Device Type managed object for every device supported by this microservice. Skips the creation if the Device Type already exists.

File: c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/service/CodecService.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+     *
+     * @param event
+     */
+    @EventListener
+    private void registerDeviceTypes(MicroserviceSubscriptionAddedEvent event) {
+        log.info("Creating device types on codec microservice subscription");
+    }
+
+    /**
+     * This method should delete a device type upon unsubscribing the codec microservice.
```
Comment (bbyreddy) [inline c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/service/CodecService.java]: - [ ] This method deletes the Device Type managed object created earlier for every device supported by this microservice. Skips the deletion if the Device Type is associated with a device.

File: c8y-lora-sdk/src/test/resources/application.properties [hunk 11]
```diff
@@ -0,0 +1,3 @@
+application.name=c8y-lora-sdk
+server.port=8071
```
Comment (bbyreddy) [inline c8y-lora-sdk/src/test/resources/application.properties]: - [ ] Is this server.port property mandated for every microservice?

File: microservice/custom-encoders/pom.xml [hunk 12]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>clients-java</artifactId>
+        <groupId>com.nsn.cumulocity.clients-java</groupId>
+        <version>1011.67.0-SNAPSHOT</version>
```
Comment (miryalarahul) [inline microservice/custom-encoders/pom.xml]: please remove this hard coded version and maintain like other pom version

File: microservice/pom.xml [hunk 13]
```diff
@@ -38,10 +38,12 @@
         <module>properties</module>
         <module>subscription</module>
         <module>logging</module>
+        <module>custom-decoders</module>
+        <module>custom-encoders</module>
```
Comment (miryalarahul) [inline microservice/pom.xml:42]: check the order is correct or not

**Review by miryalarahul — DISMISSED** [review]

**Review by ashhSAG93 — DISMISSED** [review]

**Review by bbyreddy — DISMISSED** [review]

**Review by miryalarahul — APPROVED** [review]

**Review by bbyreddy — APPROVED** [review]

**Comment by deepakgoenka** [comment]
This PR needs to be reverted as post-merge has got into a bad shape due to few recent PRs

**Pull request merged**